kinesis/kinesis.go

```diff
@@ -0,0 +1,311 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//  http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package kinesis containers the OutputPlugin which sends log records to Kinesis Stream
+package kinesis
+
+import (
+    "fmt"
+    "os"
+    "time"
+    "math/rand"
+    "errors"
+
+    "github.com/aws/amazon-kinesis-firehose-for-fluent-bit/plugins"
+    "github.com/aws/aws-sdk-go/aws"
+    "github.com/aws/aws-sdk-go/aws/awserr"
+    "github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+    "github.com/aws/aws-sdk-go/aws/endpoints"
+    "github.com/aws/aws-sdk-go/aws/session"
+    "github.com/aws/aws-sdk-go/service/kinesis"
+    fluentbit "github.com/fluent/fluent-bit-go/output"
+    jsoniter "github.com/json-iterator/go"
+    "github.com/sirupsen/logrus"
+)
+
+const (
+    partitionKeyCharset = "abcdefghijklmnopqrstuvwxyz" + "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+)
+
+var (
+    seededRand *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
Can you change `seededRand` to be a field on `OutputPlugin`?

Multiple instances of `OutputPlugin` can exist at the same time, if the user has multiple `[OUTPUT]` configurations for kinesis. With `seededRand` as a global variable, all instances of `OutputPlugin` will use the same one. This is fine for the purposes of generating random strings, but I do not think math/rand is thread safe. I'm pretty sure Fluent Bit is single threaded, but not 100% certain. So it would be safer if each `OutputPlugin` had its own `seededRand`.

Also- global variables are generally considered to be a bad practice. If its possible to avoid using them, then you should do so.

Reply:
good idea!
Now seededRandom is a filed of our OutputPlugin.